### DISCUSSION TITLE
SSO Integration with OpenID Connect but got Unauthorized

### DISCUSSION THREAD

**Discussion by xnohat** [Budibase/budibase#4065] [Help]
Hi, 

I need some help on intergrate SSO to my self-hosted Budibase.
I configured SSO with param like below
config url: https://id.domain.com/.well-known/openid-configuration
clientid: xxxx
clientsecret: xxxx
callback: http://budibase.domain.com/api/global/auth/oidc/callback

Callback Response:
http://budibase.domain.com/api/global/auth/oidc/callback?code=2c2673111968552e825f34761f7996f1648a01e2484eed3dae302b3562275aba&scope=openid%20profile%20email&state=%7B%22handle%22%3A%22qnV7WNKllQmPPUs0okbOeWiF%22%7D&session_state=BlugyKR374Nb7l-KSCfZQyCz-_3aFywd4Jkbd3YdZOA.35dffaa359bbc3d5959e948b8b8be3fd

But Budibase show Unauthorized when got callback

{"message":"Unauthorized","status":403}

**Comment by mjashanks** [comment]
Hi @xnohat - One thing I noticed about your callback URL is that it is not `https` - this is a requirement for OIDC, unless running on `localhost` for testing.

**Reply by xnohat** [reply]
Yes, I have been changed it to https, but still got {"message":"Unauthorized","status":403} :(

**Reply by mjashanks** [reply]
Sorry @xnohat  - just coming back to this.

Is your Budibase definitely running on https (sorry for the silly question)?

What SSO provider are you using?

**Reply by rrrfialho** [reply]
I´m facing the same issue. I believe it happens when budibase is behind a load balancer/reverse proxy, because the load balancer endpoint is https, and the connection from the load balancer to budibase goes through http 80 as it inside the network. This way I believe budibase doesnt recognize as https..something like this I imagine.

Do you think there is any configuration for this?

**Reply by Shyru** [reply]
Yeah, I ran into the same issue. We also have budibase behind a reverse proxy (on another host) that does https. 
Will this problem go away if we have another ssl proxy directly on the host that is running budibase via docker-compose?

**Reply by jbc-avk** [reply]
Hi @mjashanks 

I am also facing the same issue, here is the configuration : 
- budibase is running on linux lxc (ubuntu 21.10), localhost:10000 
- bb is exposed to the Internet using Caddy (in reverse proxy mode as described in the bb official doc) or haproxy+certbot.

Same result for both reverse-proxy  when using SSO (Azure AD) : {"message":"Unauthorized","status":403}

bb is working perfectly well using local accounts logon

any idea as this is blocking our tests ? the SSO is indeed a must have before we start developing our first application

many thanks

**Reply by jbc-avk** [reply]
hello @mjashanks 
looked again agt the documentation and tried many different setup but we are still facing this blocking issue
what woul be your advices ?
many thanks

**Reply by JasonBLandmark** [reply]
I'm having the same issue.  Did you find a solution?

**Reply by jbc-avk** [reply]
@JasonBLandmark  not unfortunately : despite its quality, the lack of support and this issue SSO is blocking for us so we paused the evaluation of Budibase so far

**Reply by JasonBLandmark** [reply]
I found my problem.  I was using the Secret ID form Azure and not the Value for the settings in budibase.

![image](https://user-images.githubusercontent.com/93925176/164226955-90e2a51c-78ea-419c-acb1-f03e8a4985d8.png)

After putting the value in the Field instead of the Secret ID,  the {"message":"Unauthorized","status":403} message went away and the client can now log in with Azure AD.

**Reply by man-tax** [reply]
So I'm getting the same error (403 unauthorized), but definitely have 'value' from Azure in my auth settings (double checked). Password login works for this user. https is also not the problem.

Any further ideas?

Thanks!
Sebastian

**Comment by xnohat** [comment]
Hi for anyone got this blocking issue, I have some notes in my case for your debugging

1. Use all HTTPS for either Budibase and OIDC Provider, mixed HTTPS/HTTP will blocking some redirection step of OIDC authentications
2. OIDC steps: 
	1. Client prepares an Authentication Request containing the desired request parameters.
	2. Client sends the request to the Authorization Server.
	3. Authorization Server authenticates the End-User.
	4. Authorization Server obtains End-User Consent/Authorization.
	5. Authorization Server sends the End-User back to the Client with code.
	6. Client sends the code to the Token Endpoint to receive an Access Token and ID Token in the response.
	7. Client validates the tokens and retrieves the End-User's Subject Identifier.
3. After authenticate successful, Budibase will sync User info from ODIC Provider with Budibase, it has step download Picture field from id_token (JWT encoded response from /connect/token endpoint , decode debug using tool [jwt.io](http://jwt.io/) ), please ensure that it can fetch picture from Picture field URL, or it will hang and cannot sync user with Budibase
https://github.com/Budibase/budibase/blob/develop/packages/backend-core/src/middleware/passport/third-party-common.js

Check your connection from budibase server to OIDC Provider server make sure successful, also check connection from BB to Picture URL is successful too

Source code of OIDC Authentication:
https://github.com/Budibase/budibase/blob/develop/packages/worker/src/api/controllers/global/auth.js
https://github.com/Budibase/budibase/blob/develop/packages/backend-core/src/middleware/passport/third-party-common.js
